### ISSUE TITLE
Some queries doesn't work anymore with neo4j 4.0

### ISSUE THREAD

**Issue by jonasdumas** [neo4j-graphql/neo4j-graphql-js#445]
Hi,

We try to migrate to neo4j 4.0 but we have a problem with `neo4j-graphql-js` on few queries.
Here's an example. 

**The following query generated by `neo4j-graphql-js`** is working with a `neo4j 3.5.14` DB but **doens't work anymore with `4.0.4`**

```
WITH apoc.cypher.runFirstColumn("MATCH (u:User) WHERE id(u)=toInteger($cypherParams.currentUserId) RETURN u", {offset:$offset, first:$first, cypherParams: $cypherParams}, True) AS x UNWIND x AS `user` RETURN `user` { .ttuid ,followingSpaces: [ user_followingSpaces IN apoc.cypher.runFirstColumn("MATCH (this)-[:FOLLOWS]->(:UserGroup {role: 'SUBSCRIBERS'})<-[:RIGHTS]-(space) RETURN space", {this: user, cypherParams: $cypherParams}, true) | user_followingSpaces { .ttuid , .slug }] } AS `user`
```
**The type-def who generates this query:**

```
User {
     ** various properties**
     followingSpaces: [Space] @cypher(statement:"MATCH (this)-[:FOLLOWS]->(:UserGroup {role: 'SUBSCRIBERS'})<-[:RIGHTS]-(space) RETURN space"),
}
```

**The error thrown by Cypher:**
```
Unknown variable `user_followingSpaces`.
```

Any idea why ?

**Comment by cmillsop** [comment]
I also encountered this issue.  My queries were working in 4.0.3.  I ended up rolling back my version from 4.0.4 to 4.0.3 for the time being until this is resolved or better documented.

**Comment by johnymontana** [comment]
Thanks for reporting @jonasdumas. This sounds like perhaps something related to the APOC procedure we use to handle `@cypher` directive fields. What version of APOC are you using? And did you change versions of APOC as well?

@cmillsop Could you please share an example of the type defs and GraphQL query where you're experiencing the inconsistencies between Neo4j v4.0.4 and v4.0.3? It might also be worth checking against the more recently released Neo4j v4.0.5 version as well.

**Comment by johnymontana** [comment]
OK after upgrading our integration tests to use the latest Neo4j (v4.0.5) and APOC (v4.0.0.14) I'm noticing some test failures, specifically:

```
  ✖ integration › integration › should be able to query custom cypher field returning interface type 
  ✖ integration › integration › query union type using complex fragments 
```

(previous we were testing against Neo4j v4.0.0 and APOC v4.0.0.4)

Will investigate.

Looks like both these failures are specific to interface and union types. @jonasdumas and @cmillsop are the errors you're seeing specific to interface or union types?

**Comment by jonasdumas** [comment]
Hi @johnymontana,

Thanks for invesigate.

APOC version was 4.0.0.12

Some queries are working but not the one described in the previous comment (which are not related to Union/Interface). And in general, we don't have any interface or union types in our schema.

**Comment by rbennell** [comment]
I've experienced the exact same issues with neo4j 4.0.5/ apoc 4.0.0.14, and an incredibly simple schema. Any updates on this issue?

**Comment by johnymontana** [comment]
@rbennell Would you mind sharing your code / schema so I can duplicate? I've added a specific [Neo4j v4.0.3 workflow](https://app.circleci.com/pipelines/github/neo4j-graphql/neo4j-graphql-js/38/workflows/196982f3-8f29-4754-84e6-3444a8129bda/jobs/7772) to our integration tests which is green, so for now I would recommend using Neo4j v4.0.3 if you run into this issue until we resolve the root problem.

Also, this issue doesn't seem to be present with Neo4j v4.1.0 (released today), as we have [green integration tests](https://app.circleci.com/pipelines/github/neo4j-graphql/neo4j-graphql-js/38/workflows/196982f3-8f29-4754-84e6-3444a8129bda/jobs/7771) against Neo4j v4.1.0

**Comment by rbennell** [comment]
Hi @johnymontana thanks for coming back to me. With apoc 4.0.0.14 and neo4j-enterprise 4.04/05/06 simple example of a failing query can be seen in this schema
```type Organisation {
  organisationId: ID!
  name: String!
  owns: [Owns]
  is_customer: Boolean!
  subsidiaries: [Organisation]
    @cypher(
      statement: "MATCH (this)-[:OWNS|CONTROLS*]->(o:Organisation) RETURN o"
    )
}```

{Organisation{subsidiaries{name}}} returns 
```{
  "errors": [
    {
      "message": "Unknown variable `organisation_subsidiaries`.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "Organisation"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "code": "Neo.DatabaseError.Statement.ExecutionFailed",
          "name": "Neo4jError",
          "stacktrace": [
            "Neo4jError: Unknown variable `organisation_subsidiaries`.",
            ": ",
            "    at captureStacktrace (/app/node_modules/neo4j-driver/lib/result.js:263:15)",
            "    at new Result (/app/node_modules/neo4j-driver/lib/result.js:68:19)",
            "    at newCompletedResult (/app/node_modules/neo4j-driver/lib/transaction.js:449:10)",
            "    at Object.run (/app/node_modules/neo4j-driver/lib/transaction.js:287:14)",
            "    at Transaction.run (/app/node_modules/neo4j-driver/lib/transaction.js:123:32)",
            "    at /app/node_modules/neo4j-graphql-js/dist/index.js:156:25",
            "    at TransactionExecutor._safeExecuteTransactionWork (/app/node_modules/neo4j-driver/lib/internal/transaction-executor.js:134:22)",
            "    at TransactionExecutor._executeTransactionInsidePromise (/app/node_modules/neo4j-driver/lib/internal/transaction-executor.js:122:32)",
            "    at /app/node_modules/neo4j-driver/lib/internal/transaction-executor.js:61:15",
            "    at new Promise (<anonymous>)"
          ]
        }
      }
    }
  ],
  "data": {
    "Organisation": null
  }
}```

Though I'm not certain, i think the stock create-grandstack-app fails in the same way if upgraded to these packages.
I was unable to run the app at all using neo4j 4.0.3/4.0.2, seeing the following error- 
```neo4j_1  | 2020-06-24 08:06:13.675+0000 INFO  ======== Neo4j 4.0.3 ========
neo4j_1  | 2020-06-24 08:06:13.678+0000 INFO  Starting...
neo4j_1  | 2020-06-24 08:06:22.684+0000 ERROR Failed to start Neo4j: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabaseService@412c5e8b' was successfully initialized, but failed to start. Please see the attached cause exception "Some jar procedure files are invalid, see log for details.". Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabaseService@412c5e8b' was successfully initialized, but failed to start. Please see the attached cause exception "Some jar procedure files are invalid, see log for details.".```
From what I've read, I understand this is a compatibility with this particular version of APOC.

I'm unable to test 4.1 as it's not in the docker hub yet. I'll give it a go when it's there though. For now, we've rolled our project back to 3.5.x, since create-grandstack-app uses 3.5. Are there plans to upgrade create-grandstack-app to 4.x?

**Comment by cmillsop** [comment]
Thanks for investigating.  My issue was resolved when testing with the most recent releases:
Neo4j 4.1.0
APOC 4.1.0.0
neo4j-graphql-js 2.14.3

**Comment by cmillsop** [comment]
Actually, upon further testing I am still encountering this issue.

This query fails:
`{
  Site(orderBy: name_asc) {
    _id
    name
    processVersions {
      _id
      name
      version
    }
    has_building(orderBy: name_asc) {
      _id
      name
      has_room(orderBy: name_asc) {
        _id
        name
        has_equipment(orderBy: name_asc) {
          _id
          name
        }
      }
    }
  }
}
`
With this error message:
``
2020-07-07 22:30:27.025+0000 ERROR id:12335 - 11 ms: 32520 B - bolt-session     bolt    neo4j-graphql-js/2.14.3         client/xxx.xxx.xxx.xxx:60201     server/xxx.xxx.xxx.xxx:7687> neo4j -  - MATCH (`site`:`Site`) WITH `site` ORDER BY site.name ASC RETURN `site` {_id: ID(`site`), .name ,processVersions: [ site_processVersions IN apoc.cypher.runFirstColumn("match (pv:ProcessVersion) match (this:Site) call apoc.path.subgraphNodes(this, {relationshipFilter:'HAS_ATTRIBUTE|HAS_EQUIPMENT>|HAS_ROOM>|HAS_BUILDING>|<HAS_PHASE|<HAS_OPERATION', terminatorNodes:[pv], maxLevel:10 }) yield node return node", {this: site}, true) | site_processVersions {_id: ID(`site_processVersions`),name: apoc.cypher.runFirstColumn("MATCH (this:ProcessVersion)<-[:HAS_VERSION]-(p:Process) return p.name", {this: site_processVersions}, false), .version }] ,has_building: apoc.coll.sortMulti([(`site`)-[:`HAS_BUILDING`]->(`site_has_building`:`Building`) | `site_has_building` {_id: ID(`site_has_building`), .name ,has_room: apoc.coll.sortMulti([(`site_has_building`)-[:`HAS_ROOM`]->(`site_has_building_has_room`:`Room`) | `site_has_building_has_room` {_id: ID(`site_has_building_has_room`), .name ,has_equipment: apoc.coll.sortMulti([(`site_has_building_has_room`)-[:`HAS_EQUIPMENT`]->(`site_has_building_has_room_has_equipment`:`Equipment`) | `site_has_building_has_room_has_equipment` {_id: ID(`site_has_building_has_room_has_equipment`), .name }], ['^name']) }], ['^name']) }], ['^name']) } AS `site` - {1_orderBy: 'name_asc', 2_orderBy: 'name_asc', offset: 0.0, first: -1.0, 3_orderBy: 'name_asc'} - runtime=pipelined - {} - Unknown variable `site_processVersions`.
``

However, if I modify the query by removing any of the properties beneath Site or its sub-entities or if I remove any of the orderBy's then the query resolves successfully.  Here is an example log message from a success with no "_id" property on Site:

``
20-07-07 22:34:38.884+0000 INFO  id:12424 - 8 ms: 33096 B - bolt-session      bolt    neo4j-graphql-js/2.14.3         client/xxx.xxx.xxx.xxx:60201     server/xxx.xxx.xxx.xxx:7687> neo4j -  - MATCH (`site`:`Site`) WITH `site` ORDER BY site.name ASC RETURN `site` { .name ,processVersions: [ site_processVersions IN apoc.cypher.runFirstColumn("match (pv:ProcessVersion) match (this:Site) call apoc.path.subgraphNodes(this, {relationshipFilter:'HAS_ATTRIBUTE|HAS_EQUIPMENT>|HAS_ROOM>|HAS_BUILDING>|<HAS_PHASE|<HAS_OPERATION', terminatorNodes:[pv], maxLevel:10 }) yield node return node", {this: site}, true) | site_processVersions {_id: ID(`site_processVersions`),name: apoc.cypher.runFirstColumn("MATCH (this:ProcessVersion)<-[:HAS_VERSION]-(p:Process) return p.name", {this: site_processVersions}, false), .version }] ,has_building: apoc.coll.sortMulti([(`site`)-[:`HAS_BUILDING`]->(`site_has_building`:`Building`) | `site_has_building` {_id: ID(`site_has_building`), .name ,has_room: apoc.coll.sortMulti([(`site_has_building`)-[:`HAS_ROOM`]->(`site_has_building_has_room`:`Room`) | `site_has_building_has_room` {_id: ID(`site_has_building_has_room`), .name ,has_equipment: apoc.coll.sortMulti([(`site_has_building_has_room`)-[:`HAS_EQUIPMENT`]->(`site_has_building_has_room_has_equipment`:`Equipment`) | `site_has_building_has_room_has_equipment` {_id: ID(`site_has_building_has_room_has_equipment`), .name }], ['^name']) }], ['^name']) }], ['^name']) } AS `site` - {1_orderBy: 'name_asc', 2_orderBy: 'name_asc', offset: 0.0, first: -1.0, 3_orderBy: 'name_asc'} - runtime=pipelined - {}
``

**Comment by benjamin-rood** [comment]
We are having a similar issue, following our upgrade from Neo4j 3.5 to 4.1.0

Debug logging ON:
```
2020-07-16T03:59:58.931Z neo4j-graphql-js WITH apoc.cypher.runFirstColumn("MATCH (reseller:Reseller)
WHERE reseller.guid IN $cypherParams.nodes
RETURN reseller", {offset:$offset, first:$first, cypherParams: $cypherParams}, True) AS x UNWIND x AS `reseller` RETURN `reseller` {properties: head([ reseller_properties IN apoc.cypher.runFirstColumn("MATCH (this)-[r:HAS_PROPERTIES]->(rp:ResellerProperties)
WHERE r.validUntil IS NULL AND r.validFrom IS NOT NULL
RETURN rp ORDER BY r.validFrom DESC LIMIT 1", {this: reseller, cypherParams: $cypherParams}, true) | reseller_properties { .guid , .name }]) , .guid , .accountID , .deleted } AS `reseller`
2020-07-16T03:59:58.932Z neo4j-graphql-js {
"offset": 0,
"first": -1,
"cypherParams": {
"nodes": [
"43eb94f1-f1f8-4e3c-b6eb-db7ba930352c"
]
}
}
{"level":50,"time":1594871998948,"pid":17,"hostname":"graphql-neo4j-service-6f8f848569-zchdc","name":"graphql-neo4j-mw","msg":"GraphQLError =  {\"message\":\"Unknown variable `reseller_properties`.\",\"locations\":[{\"line\":2,\"column\":3}],\"path\":[\"resellers\"],\"extensions\":{\"code\":\"INTERNAL_SERVER_ERROR\",\"exception\":{\"code\":\"Neo.DatabaseError.Statement.Executition.Failure
(etc)
```

While copying and pasting this EXACT query into `cypher-shell` will work fine:
```
neo4j@staging> :param cypherParams=>{nodes: ["43eb94f1-f1f8-4e3c-b6eb-db7ba930352c"]}
neo4j@staging> :param offset=>0
neo4j@staging> :param first=>-1
neo4j@staging> WITH apoc.cypher.runFirstColumn("MATCH (reseller:Reseller)
WHERE reseller.guid IN $cypherParams.nodes
RETURN reseller", {offset:$offset, first:$first, cypherParams: $cypherParams}, True) AS x UNWIND x AS `reseller` RETURN `reseller` {properties: head([ reseller_properties IN apoc.cypher.runFirstColumn("MATCH (this)-[r:HAS_PROPERTIES]->(rp:ResellerProperties)
WHERE r.validUntil IS NULL AND r.validFrom IS NOT NULL
RETURN rp ORDER BY r.validFrom DESC LIMIT 1", {this: reseller, cypherParams: $cypherParams}, true) | reseller_properties { .guid , .name }]) , .guid , .accountID , .deleted } AS `reseller`;
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| reseller                                                                                                                                                                                 |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {guid: "43eb94f1-f1f8-4e3c-b6eb-db7ba930352c", accountID: "2136587078", deleted: NULL, properties: {name: "Vadaxchange Testing Reseller", guid: "d7b99236-50fe-4819-9d1d-4326c74303ab"}} |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Which is the correct and expected output.

We cannot replicate this locally, this is only ccurring in a clustered environment. So, it sounds like it has nothing to do with `neo4j-graphql-js`, and yet...

**Comment by gwvandesteeg** [comment]
We've reverted the code used by @benjamin-rood and myself to using the 2.14.0 version and the problem appears to have gone away.  This is also the last release which built correctly and had all tests pass according to the CircleCI tests.

**Comment by gwvandesteeg** [comment]
Some more details to help identify the problem, a failing query:
```
MATCH (`reseller`:`Reseller` {guid:$guid}) RETURN `reseller` {properties: head([ reseller_properties IN apoc.cypher.runFirstColumn("MATCH (this)-[r:HAS_PROPERTIES]->(rp:ResellerProperties)
WHERE r.validUntil IS NULL AND r.validFrom IS NOT NULL
RETURN rp ORDER BY r.validFrom DESC LIMIT 1", {this: reseller}, true) | reseller_properties { .guid , .name }]) , .guid , .accountID , .deleted } AS `reseller`
```
Inputs:
```
{
"offset": 0,
"first": -1,
"guid": "43eb94f1-f1f8-4e3c-b6eb-db7ba930352c"
}
```
The related schema (excerpts) for that would be:
```
type Reseller {
  guid: String!
  accountID: String!
  deleted: Boolean
  properties: ResellerProperties
    @cypher(
      statement: """
      MATCH (this)-[r:HAS_PROPERTIES]->(rp:ResellerProperties)
      WHERE r.validUntil IS NULL AND r.validFrom IS NOT NULL
      RETURN rp ORDER BY r.validFrom DESC LIMIT 1
      """
    )
}

type ResellerProperties {
  guid: String!
  name: String!
  description: String
}
```
And the cluster overview.. we're running the queries against the `staging` database:
```
neo4j@neo4j> CALL dbms.cluster.overview();
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                     | addresses                                                                                                                                                    | databases                                                  | groups |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| "c53ed1d6-9b7f-4ea0-854c-7497b881919a" | ["bolt://neo4j410-neo4j-core-1.neo4j410-neo4j.default.svc.cluster.local:7687", "http://neo4j410-neo4j-core-1.neo4j410-neo4j.default.svc.cluster.local:7474"] | {neo4j: "LEADER", system: "FOLLOWER", staging: "FOLLOWER"} | []     |
| "dfbfdf49-899d-4279-88b9-eaf344515564" | ["bolt://neo4j410-neo4j-core-2.neo4j410-neo4j.default.svc.cluster.local:7687", "http://neo4j410-neo4j-core-2.neo4j410-neo4j.default.svc.cluster.local:7474"] | {neo4j: "FOLLOWER", system: "LEADER", staging: "FOLLOWER"} | []     |
| "edebf318-ba3c-4c35-a543-690b5a8aae14" | ["bolt://neo4j410-neo4j-core-0.neo4j410-neo4j.default.svc.cluster.local:7687", "http://neo4j410-neo4j-core-0.neo4j410-neo4j.default.svc.cluster.local:7474"] | {neo4j: "FOLLOWER", system: "FOLLOWER", staging: "LEADER"} | []     |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
if the query runs against the LEADER node, it works, if run against a FOLLOWER node it does not work. On a FOLLOWER node we get:
```2020-07-17 04:21:58.335+0000 ERROR Client triggered an unexpected error [Neo.DatabaseError.Statement.ExecutionFailed]: Unknown variable `reseller_properties`., reference 16424391-7417-4074-a005-3087d6580b0e.```

**Comment by gwvandesteeg** [comment]
Digging further and the cause of the problem in the graphql side of it is an interplay between nested fragments, some work.. some don't.  If we do the query with no nesting it works.

This is the original query with fragments that dies, where we have nesting fragments:
```
fragment ResellerPropertiesFields on ResellerProperties {
  guid
  name
  __typename
}

fragment ResellerFields on Reseller {
  guid
  accountID
  deleted
  __typename
}

fragment ResellerItemFields on Reseller {
    ...ResellerFields
    properties {
      ...ResellerPropertiesFields
    }  
}

query {
  resellers {
	...ResellerItemFields
  }
}
```
This one with no nested fragments works:
```
fragment ResellerPropertiesFields on ResellerProperties {
  guid
  name
  __typename
}

fragment ResellerFields on Reseller {
  guid
  accountID
  deleted
  __typename
}

query {
  resellers {
    ...ResellerFields
    properties {
      ...ResellerPropertiesFields
    }
  }
}
```

And this one with nested fragments works:
```
fragment ResellerPropertiesFields on ResellerProperties {
  guid
  name
  __typename
}

fragment ResellerItemFields on Reseller {
  guid
  accountID
  deleted
  properties {
    ...ResellerPropertiesFields
  }
}

{
  resellers {
    ...ResellerItemFields
  }
}
```

**Comment by gwvandesteeg** [comment]
And nope it wasn't fragments.. it's the ordering of the attributes and relationships being retrieved:
This breaks:
```
query {
  resellers (deleted: null) {
    properties {
			guid
      name
      __typename
    }
    guid
    accountID
    deleted
    __typename
  }
}
```
This works:
```
query {
  resellers (deleted: null) {
    guid
    accountID
    deleted
    __typename
    properties {
			guid
      name
      __typename
    }
  }
}
```

**Comment by gwvandesteeg** [comment]
What the... it gets better.. the working query above.. worked when i posted it, the exact same query without changing it running approx 30 minutes later.. doesn't work..  The only difference is a) time, and b) other people running queries against the graphql and neo4j servers.

**Comment by gwvandesteeg** [comment]
Narrowed it down some more, when running the generated cypher query via the cypher-shell and repeating the same query over and over again periodically reproduces the error state.

```
1 row available after 1 ms, consumed after another 2 ms
neo4j@staging> WITH apoc.cypher.runFirstColumn("MATCH (reseller:Reseller) RETURN reseller", {offset:$offset, first:$first, cypherParams: $cypherParams}, True) AS x
UNWIND x AS `reseller` RETURN `reseller` {
.guid , .accountID , .deleted ,properties: head([ reseller_properties IN apoc.cypher.runFirstColumn("MATCH (this)-[r:HAS_PROPERTIES]->(rp:ResellerProperties) WHERE r.validUntil IS NULL AND r.validFrom IS NOT NULL RETURN rp ORDER BY r.validFrom DESC LIMIT 1", {this: reseller, cypherParams: $cypherParams}, true) | reseller_properties { .guid, .name }])
}
AS `reseller`;
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| reseller                                                                                                                                                                                 |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {guid: "b4abe751-d8d8-4560-bd19-d70acb1ae7a0", accountID: "7272480929", deleted: NULL, properties: {name: "Vadaxchange Testing Reseller", guid: "2f587e3e-8f1c-455d-bca9-b3dc7a9b3a98"}} |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row available after 1 ms, consumed after another 1 ms
neo4j@staging> WITH apoc.cypher.runFirstColumn("MATCH (reseller:Reseller) RETURN reseller", {offset:$offset, first:$first, cypherParams: $cypherParams}, True) AS x
UNWIND x AS `reseller` RETURN `reseller` {
.guid , .accountID , .deleted ,properties: head([ reseller_properties IN apoc.cypher.runFirstColumn("MATCH (this)-[r:HAS_PROPERTIES]->(rp:ResellerProperties) WHERE r.validUntil IS NULL AND r.validFrom IS NOT NULL RETURN rp ORDER BY r.validFrom DESC LIMIT 1", {this: reseller, cypherParams: $cypherParams}, true) | reseller_properties { .guid, .name }])
}
AS `reseller`;
Unknown variable `reseller_properties`.
neo4j@staging> :params
:param cypherParams => {nodes: ['b4abe751-d8d8-4560-bd19-d70acb1ae7a0']}
:param first        => -1
:param offset       => 0
```
The above error is the 10th iteration of running the exact same query against the standalone enterprise instance using apoc 4.1.0.1 and neo4j 4.1.0.0 .

**Comment by gwvandesteeg** [comment]
Tried various versions of APOC on neo4j 4.1.0, 4.0.5, 4.0.4 and every single one of them shows the problem within 10 consecutive runs of the query above.  100% ability to reproduce it on a local docker image as well.  This is definitely a neo4j 4.x problem.
In the process of stepping back neo4j versions i've stepped back to neo4j 3.5.14 and this is the first time i've not been able to cause the problem in 200 consecutive runs of the same query.

**Comment by cmillsop** [comment]
I ended up rolling back to Neo4j v4.0.3 and APOC v4.0.0.15.  I can confirm that these versions don't produce the bug being discussed here.

**Comment by gwvandesteeg** [comment]
Nope, they don't produce this bug, they produce another one instead when using CASE statements inside @cypher directives.

**Comment by johnymontana** [comment]
I believe this issue has now been resolved in Neo4j v4.0.7 and v4.1.1 Specifically, as mentioned in [the 4.0 changelog:](https://github.com/neo4j/neo4j/wiki/Neo4j-4.0-changelog#cypher) and [the 4.1 changelog:](https://github.com/neo4j/neo4j/wiki/Neo4j-4.1-changelog#cypher)

> Fixed a bug where map projection inside a list or pattern comprehension would fail with Unknown variable error.

Our previous failing integration tests that were triggering this issue are now passing which I'm hoping confirms the fix.

If you're able to upgrade to those versions and confirm the fix that'd be appreciated.

**Comment by gwvandesteeg** [comment]
Confirmed resolved in v4.07
